library/src/main/scala/scala/scalajs/js/Thenable.scala

```diff
@@ -80,4 +84,23 @@ object Thenable {
     implicit def thenable2future[A](p: Thenable[A]): Future[A] =
       p.toFuture
   }
+
+  sealed trait Returning[A, B]
+
+  trait Returning1 {
+    implicit def default[A]: Returning[A, A] = new Returning[A, A] {}
+  }
+
+  object Returning extends Returning1 {
+
+    implicit def flattened[A, T[X] <: Thenable[X]]: Returning[T[A], A] =
+      new Returning[T[A], A] {}
```

Comment:
I'm pretty sure this can still be broken. Consider:
```scala
def mapP[A, B](p: js.Promise[A], f: A => B): js.Promise[B] =
  p.`then`(x => f(x))

val p = js.Promise(5)
val p2: js.Promise[js.Promise[Int]] = mapP(p, x => js.Promise(x))
```